### DISCUSSION TITLE
How to dedupe tailwind classes in bodyAttrs?

### DISCUSSION THREAD

**Discussion by chrisjansky** [unjs/unhead#431] [Q&A]
Hi,

I am using `unhead` as part of `nuxt@3`. Is there any way to dedupe tailwind classes when `useHead()` is used in both `app.vue` and in `pages/`?

Example:

> app.vue
```
useHead({
  bodyAttrs: {
    class: "bg-red-500",
  },
})
```

> pages/index.vue
```
useHead({
  bodyAttrs: {
    class: "bg-green-500",
  },
})
```

Ideally I would dedupe them with something like [tw-merge](https://www.npmjs.com/package/tailwind-merge) to get `class="bg-green-500"` only in the final output.

Basically any approach for specifying __default__ `<body>` classes and __overrides__ for pages would be great.

Thanks :pray:

**Comment by harlan-zw** [comment]
Hi, thanks for the question. The simplest way to solve this would be to conditionally add the class in your app.vue.

```ts
const route = useRoute()
useHead({
  bodyAttrs: {
    class: {
      "bg-red-500": () => route.path !== '/'
    }
  },
})
```

There is a more complicated way to solve this in where we can use the Unhead hooks to modify the resolved tags using tw-merge, this is a rough untested code sample of how you might do that

```ts
const head = injectHead()
head.hooks.hook('tags:afterResolve', (ctx) => {
  // collect all body classes
  const bodyClasses = ctx.tags.filter((tag) => tag.tag === 'bodyAttrs').flatMap((tag) => tag.props.class)
  // apply tw-merge
  const twClasses = twMerge(bodyClasses)
  // filter out non-merged classes
  for (const tag of ctx.tags) {
    if (tag.tag === 'bodyAttrs') {
      tag.props.class = twClasses
    }
  }
})
```

**Reply by chrisjansky** [reply]
Hi, @harlan-zw, thanks for the quick reply (and HNY!)

The second snippet works great, thank you.

One follow-up question: where would be the place to put it outside of `app.vue` for a more modularized structure? I tried creating a new `nuxt` plugin at `app/plugins/plugin-unhead.ts` with the following:

```javascript
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hooks.hook("app:created", (app) => {

    // same as before
    const head = injectHead()
    head.hooks.hook("tags:resolve", (ctx) => {
      // collect all body classes
      const bodyClasses = ctx.tags.filter((tag) => tag.tag === "bodyAttrs").flatMap((tag) => tag.props.class)
      // apply tw-merge
      const twClasses = cn(bodyClasses)
      // filter out non-merged classes
      for (const tag of ctx.tags) {
        if (tag.tag === "bodyAttrs") {
          tag.props.class = twClasses
        }
      }
    })

  })
})
```

But that doesn’t work properly – it runs, but not at the right time (supposedly) so the tags are not deduped, only concatenated. I am not that fluent with `nuxt` hooks, but with some debugging I got the idea that it needs to run in SSR (at runtime in both server and client), but with this `app:created` hook it only runs in client for me.

Nuxt docs lists `app:created` as the earliest SSR hook (AFAIK): https://nuxt.com/docs/api/advanced/hooks so I wonder if there’s a different hook to use or perhaps not a plugin altogether.

Any pointers appreciated!

TL;DR where to put the second snippet outside of `app.vue` to get identical behavior?

**Comment by chrisjansky** [comment]
Hi @harlan-zw, sorry to necro this issue – the newly released unhead v2 as part of nuxt 3.16.2 breaks the `injectHead()` hook you have written: bodyClasses props.class return `''`.

Could you please provide any guidance how to make this behaviour work in the new version? Thanks 🙏
